File: backends/java/zappy-application/src/main/java/com/zappymart/application/accounts/RegisterCustomer.java
package com.zappymart.application.accounts;

import com.zappymart.application.Visitor;
import com.zappymart.application.ports.Clock;
import com.zappymart.application.ports.CustomerRepository;
import com.zappymart.application.ports.IdentifierGenerator;
import com.zappymart.application.ports.PasswordHasher;
import com.zappymart.application.ports.RateLimiter;
import com.zappymart.application.ports.UnitOfWork;
import com.zappymart.domain.accounts.Customer;
import com.zappymart.domain.accounts.PasswordPolicy;
import com.zappymart.domain.shared.EmailAddress;
import com.zappymart.domain.shared.Result;
import com.zappymart.domain.shared.UserErrorCode;

import java.util.Optional;

public final class RegisterCustomer {

    private final UnitOfWork unitOfWork;
    private final CustomerRepository customerRepository;
    private final PasswordHasher passwordHasher;
    private final SessionOpening sessionOpening;
    private final AnonymousHandover anonymousHandover;
    private final RateLimiter rateLimiter;
    private final IdentifierGenerator identifierGenerator;
    private final Clock clock;

    public RegisterCustomer(UnitOfWork unitOfWork, CustomerRepository customerRepository,
                            PasswordHasher passwordHasher, SessionOpening sessionOpening,
                            AnonymousHandover anonymousHandover, RateLimiter rateLimiter,
                            IdentifierGenerator identifierGenerator, Clock clock) {
        this.unitOfWork = unitOfWork;
        this.customerRepository = customerRepository;
        this.passwordHasher = passwordHasher;
        this.sessionOpening = sessionOpening;
        this.anonymousHandover = anonymousHandover;
        this.rateLimiter = rateLimiter;
        this.identifierGenerator = identifierGenerator;
        this.clock = clock;
    }

    public Result<Authentication> execute(Visitor visitor, String typedEmail, String name, String password) {
        Optional<EmailAddress> email = EmailAddress.parse(typedEmail);
        if (email.isEmpty()) {
            return Result.refuse(UserErrorCode.EMAIL_INVALID,
                    "That is not an email address.", "input.email");
        }
        if (!rateLimiter.allows("register:" + email.get().value())) {
            return Result.refuse(UserErrorCode.RATE_LIMITED,
                    "Too many attempts in a short time. Wait a moment and try again.");
        }
        Result<String> checkedPassword = PasswordPolicy.check(password);
        if (!checkedPassword.succeeded()) {
            return checkedPassword.carryRefusal();
        }
        return unitOfWork.inTransaction(() -> {
            if (customerRepository.byEmail(email.get()).isPresent()) {
                return Result.refuse(UserErrorCode.EMAIL_TAKEN,
                        "A customer with that email address is already registered.", "input.email");
            }
            Customer customer = customerRepository.save(new Customer(identifierGenerator.next(), email.get(),
                    name, passwordHasher.hash(password), clock.now()));
            anonymousHandover.toCustomer(visitor, customer.id());
            return Result.of(sessionOpening.openFor(customer, visitor.device()));
        });
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/accounts/RegisterCustomer.kt
package nl.zappymart.application.accounts

import nl.zappymart.application.Visitor
import nl.zappymart.application.cart.VisitorCart
import nl.zappymart.application.ports.Clock
import nl.zappymart.application.ports.CustomerRepository
import nl.zappymart.application.ports.IdentifierFactory
import nl.zappymart.application.ports.PasswordHasher
import nl.zappymart.application.ports.RateLimiter
import nl.zappymart.application.ports.UnitOfWork
import nl.zappymart.domain.accounts.Customer
import nl.zappymart.domain.accounts.PasswordPolicy
import nl.zappymart.domain.shared.EmailAddress
import nl.zappymart.domain.shared.Result
import nl.zappymart.domain.shared.UserErrorCode
import nl.zappymart.domain.shared.refusal

data class RegistrationRequest(val email: String, val name: String, val password: String, val device: String)

class RegisterCustomer(
    private val customers: CustomerRepository,
    private val passwords: PasswordHasher,
    private val identifiers: IdentifierFactory,
    private val clock: Clock,
    private val signIn: SignIn,
    private val visitorCart: VisitorCart,
    private val wishlistOwner: WishlistOwner,
    private val rateLimiter: RateLimiter,
    private val unitOfWork: UnitOfWork,
) {

    fun execute(visitor: Visitor, request: RegistrationRequest): Result<Authentication> {
        if (!rateLimiter.allows("register:${request.email.trim().lowercase()}")) {
            return refusal(UserErrorCode.RATE_LIMITED, "Too many attempts. Please wait and try again.")
        }
        val email = when (val parsed = EmailAddress.of(request.email)) {
            is Result.Refused -> return parsed
            is Result.Success -> parsed.value
        }
        val policy = PasswordPolicy.check(request.password)
        if (policy is Result.Refused) {
            return policy
        }
        return unitOfWork.execute {
            if (customers.findByEmail(email) != null) {
                refusal(UserErrorCode.EMAIL_TAKEN, "That email address is already registered.", "input.email")
            } else {
                val customer = customers.save(
                    Customer(
                        id = identifiers.next(),
                        email = email,
                        name = request.name.trim(),
                        passwordHash = passwords.hash(request.password),
                        createdAt = clock.moment(),
                    ),
                )
                wishlistOwner.moveToCustomer(visitor, customer.id)
                visitorCart.moveToCustomer(visitor, customer.id)
                Result.Success(signIn.start(customer, request.device))
            }
        }
    }
}
